File: backend/src/controllers/ai.controller.ts
// AI controller
// Handles AI-powered search and recommendation requests

import { Response, NextFunction } from 'express';
import { AIService } from '../services/ai.service.js';
import { ApiResponseHelper } from '../utils/responses.js';
import { AuthRequest } from '../types/index.js';

export class AIController {
  /**
   * AI-powered search
   * POST /api/ai/search
   */
  static async search(req: AuthRequest, res: Response, next: NextFunction) {
    try {
      const { query } = req.body;
      const limit = Number(req.query.limit) || 20;

      const products = await AIService.searchProducts(query, limit);

      return ApiResponseHelper.success(res, {
        query,
        parsed: AIService.parseSearchQuery(query),
        results: products,
      });
    } catch (error) {
      next(error);
    }
  }

  /**
   * Get similar products
   * GET /api/ai/recommend/:productId
   */
  static async getSimilarProducts(req: AuthRequest, res: Response, next: NextFunction) {
    try {
      const { productId } = req.params;
      const limit = Number(req.query.limit) || 5;

      const products = await AIService.getSimilarProducts(productId, limit);

      return ApiResponseHelper.success(res, products);
    } catch (error) {
      next(error);
    }
  }

  /**
   * Get personalized recommendations
   * GET /api/ai/recommend/user
   */
  static async getPersonalizedRecommendations(req: AuthRequest, res: Response, next: NextFunction) {
    try {
      if (!req.user) {
        // Return popular products for unauthenticated users
        const limit = Number(req.query.limit) || 5;
        const products = await AIService.getPopularProducts(limit);
        return ApiResponseHelper.success(res, products);
      }

      const limit = Number(req.query.limit) || 5;
      const products = await AIService.getPersonalizedRecommendations(req.user.id, limit);

      return ApiResponseHelper.success(res, products);
    } catch (error) {
      next(error);
    }
  }

  /**
   * Get popular/trending products
   * GET /api/ai/popular
   */
  static async getPopularProducts(req: AuthRequest, res: Response, next: NextFunction) {
    try {
      const limit = Number(req.query.limit) || 10;
      const products = await AIService.getPopularProducts(limit);

      return ApiResponseHelper.success(res, products);
    } catch (error) {
      next(error);
    }
  }
}

File: backend-java/src/main/java/com/ecommerce/mini/controller/AIController.java
package com.ecommerce.mini.controller;

import com.ecommerce.mini.dto.AISearchRequest;
import com.ecommerce.mini.dto.AISearchResponse;
import com.ecommerce.mini.dto.ProductResponse;
import com.ecommerce.mini.service.AIService;
import jakarta.validation.Valid;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

import java.util.List;

/**
 * REST controller for AI-powered features
 * Provides intelligent search and recommendations
 */
@RestController
@RequestMapping("/ai")
@RequiredArgsConstructor
public class AIController {

    private final AIService aiService;

    /**
     * AI-powered product search
     * POST /api/ai/search
     * Supports natural language queries like:
     * - "bluetooth headphones under 100"
     * - "gaming mouse between 50 and 150"
     * - "best rated keyboards"
     */
    @PostMapping("/search")
    public ResponseEntity<AISearchResponse> aiSearch(@Valid @RequestBody AISearchRequest request) {
        List<ProductResponse> results = aiService.aiSearch(request.getQuery());

        AISearchResponse response = AISearchResponse.builder()
                .results(results)
                .query(request.getQuery())
                .count(results.size())
                .build();

        return ResponseEntity.ok(response);
    }

    /**
     * Get similar products based on a given product
     * GET /api/ai/recommend/{productId}
     * Uses TF-IDF cosine similarity on title and tags
     */
    @GetMapping("/recommend/{productId}")
    public ResponseEntity<List<ProductResponse>> getSimilarProducts(@PathVariable String productId) {
        List<ProductResponse> recommendations = aiService.getSimilarProducts(productId);
        return ResponseEntity.ok(recommendations);
    }

    /**
     * Get personalized recommendations for the current user
     * GET /api/ai/recommend/user
     * Based on user behavior and collaborative filtering
     */
    @GetMapping("/recommend/user")
    public ResponseEntity<List<ProductResponse>> getPersonalizedRecommendations() {
        // For now, returns popular products
        // In future: use actual user ID from JWT token
        List<ProductResponse> recommendations = aiService.getPersonalizedRecommendations(null);
        return ResponseEntity.ok(recommendations);
    }
}
